Stats views widget

planarvoid opened the pull request:
This PR adds a stats views widget in two modes - Light and Dark. The widget contains a list of the last 7 days with the change from the previous day and the views per day. When you decrease the Widget width, the "Change" column should disappear. 
The items are clickable but nothing happens yet on click. I'll include that in another PR.

To test:
* Add the Stats widget
* Select any configuration (site/color)
* The widget is created and shows the data for the last 7 days for the selected site
* The widget shows the selected color
* Resize the widget to smaller size
* The "Change" column disappears
* Repeat with different parameters

<img width="264" alt="Screenshot 2019-05-28 at 15 48 42" src="https://user-images.githubusercontent.com/1079756/58483219-335bab80-8160-11e9-8eee-3ed622150dc4.png">
<img width="261" alt="Screenshot 2019-05-28 at 15 48 52" src="https://user-images.githubusercontent.com/1079756/58483220-33f44200-8160-11e9-9fd8-089b262ac785.png">
<img width="195" alt="Screenshot 2019-05-28 at 15 49 14" src="https://user-images.githubusercontent.com/1079756/58483221-33f44200-8160-11e9-8909-afffb4cb1b6a.png">
<img width="195" alt="Screenshot 2019-05-28 at 15 49 21" src="https://user-images.githubusercontent.com/1079756/58483222-33f44200-8160-11e9-86ac-2ac5b29e760a.png">

Update release notes:

- [x] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

planarvoid committed:
First widget version

planarvoid committed:
Fix the missing icons

planarvoid committed:
Change colors from grey to gray

planarvoid committed:
Refactoring

planarvoid committed:
Add tests

SylvesterWilmott commented:
Looks great @planarvoid and works as expected. Thanks!

planarvoid committed:
Merge branch 'feature/add_view_mode_selector' into feature/stats_views_widget

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/widget/StatsViewsWidget.kt:
```diff
@@ -7,59 +7,122 @@ import android.content.Intent
-        if (context != null && appWidgetManager != null) {
-            updateAppWidget(context, appWidgetManager, appWidgetId)
+        if (context != null) {
+            // See the dimensions and
```
Looks like the rest of the comment is gone :)

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/widget/StatsViewsWidget.kt:
```diff
@@ -7,59 +7,122 @@ import android.content.Intent
+            // See the dimensions and
+            val options = appWidgetManager.getAppWidgetOptions(appWidgetId)
+
+            // Get min width and height.
```
Looks like we not using height when updating the widget, so we can remove it from comment.

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/widget/StatsViewsWidget.kt:
```diff
@@ -7,59 +7,122 @@ import android.content.Intent
+            val siteIconUrl = siteStore.getSiteBySiteId(appPrefsWrapper.getAppWidgetSiteId(appWidgetId))?.iconUrl
+            val awt = AppWidgetTarget(context, id.widget_site_icon, views, appWidgetId)
+            imageManager.load(awt, context, ICON, siteIconUrl ?: "")
             val svcIntent = Intent(context, WidgetService::class.java)
```
What is scv in this context?

planarvoid replied:
I've renamed it to `listIntent`

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/widget/ViewsWidgetListViewModel.kt:
```diff
@@ -0,0 +1,106 @@
+
+const val LIST_ITEM_COUNT = 7
+
+class ViewsListViewModel
```
To stay consistent with the rest of the filenames, maybe we should call this class something like `ViewsWidgetListViewModel`. WDYT?

planarvoid replied:
Sounds good 👍

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/widget/ViewsWidgetListViewModel.kt:
```diff
@@ -0,0 +1,106 @@
+        siteId?.let {
+            val site = siteStore.getSiteBySiteId(it)
+            runBlocking {
+                visitsAndViewsStore.fetchVisits(site, DAYS, Top(LIST_ITEM_COUNT + 1), Date())
```
If we delete the site used in widget from the app, `site` will be `null`, and app will crash around here.

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/widget/ViewsWidgetListProvider.kt:
```diff
@@ -0,0 +1,78 @@
+        val uiModel = viewModel.data[position]
+        val rv = RemoteViews(context.packageName, uiModel.layout)
+        rv.setTextViewText(R.id.period, uiModel.key)
+        if (uiModel.isNeutralChangeVisible) {
```
I assume only one type of change could be visible at the time, so I wonder, maybe we could simplify this part by passing enum to `uiModel ` and use it (inside `when`, for example) to pick the view id?

planarvoid replied:
That's true but neither can be shown (if there is no previous item). This would mean we'd need some extra logic that's now in the ViewModel. I think I can create a new enum to handle these 4 usecases though 👍

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/widget/StatsViewsWidget.kt:
```diff
@@ -7,59 +7,122 @@ import android.content.Intent
+                    appPrefsWrapper,
+                    siteStore,
+                    imageManager,
+                    minWidth > 250
```
`250` is used in a couple of places - should we extract it to a resource or constant?

khaykov reviewed (commented):
Apart from the crash when site used with widget is deleted it works as expected! I left a couple of minor comments about the code :)

khaykov commented:
Also, wanted to make sure if the target branch is correct.

planarvoid committed:
Address feedback on the PR

planarvoid commented:
thanks for the review @khaykov ! I've addressed all the comments. Do you think we should somehow handle the use case when the site is missing? Something like an "Error" view. I don't think this will be that common.

planarvoid committed:
Merge branch 'feature/stats_widgets_readonly' into feature/stats_views_widget

nbradbury commented:
Just a friendly reminder to please make sure to manually merge `develop` into this branch before you merge this PR, as `develop` now contains the AndroidX migration. The automatic merge can succeed but still break the build. Better to find out now before it breaks `develop`. Thanks!

khaykov commented:
@planarvoid Thanks for the fixes. Error view sounds good. It can be used for other errors too, like when you add a widget when having no internet connection, for a site without cached views data. Currently, the widget will just remain blank until the device is restarted (btw, at what time the content of the widget is refreshed?)

planarvoid committed:
Merge branch 'feature/stats_widgets_readonly' into feature/stats_views_widget

planarvoid commented:
@khaykov let's do the error and the loading view in another PR (since this one is already pretty massive). We should definitely have that. I've just set the refresh rate to 10 mins (600000ms in the widget settings). It used to be more.

planarvoid committed:
Set widget refresh rate to 10 mins

khaykov reviewed (approved):
Sure, let's do as part of another issue :) This one looks good 👍
